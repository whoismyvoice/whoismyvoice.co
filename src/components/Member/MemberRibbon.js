import React from 'react';

// Component
import BaseComponent from '../BaseComponent';

// Styles
import style from './../../styles/MemberRibbon.scss';

class MemberRibbon extends BaseComponent {
  render() {
    let {name, state, party} = this.props;

    return <div className="ribbon-wrapper">
      <h1 className="ribbon">
        <strong className="ribbon-content">
    	     ({party} - {state}) {name}
        </strong>
      </h1>
    </div>;
  }
};

MemberRibbon.propTypes = {
  name: React.PropTypes.string,
  state: React.PropTypes.string,
  party: React.PropTypes.string
};

export default MemberRibbon;
