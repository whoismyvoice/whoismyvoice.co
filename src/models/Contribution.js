// @flow

import PropTypes from 'prop-types';

/**
 * Defines the type alias for a `Contribution` record.
 */
export type ContributionRecord = {
  amount: number,
  legislatorId: string,
  organization: string,
};

/**
 * Defines the `PropTypes` validation structure for a `Contribution` record.
 */
export const PropType = PropTypes.shape({
  amount: PropTypes.number.isRequired,
  legislatorId: PropTypes.string.isRequired,
  organization: PropTypes.string.isRequired,
});
