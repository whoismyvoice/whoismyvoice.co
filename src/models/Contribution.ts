/**
 * Defines the type alias for a `Contribution` record.
 */
export interface Contribution {
  amount: number;
  legislatorId: string;
  organization: string;
}
