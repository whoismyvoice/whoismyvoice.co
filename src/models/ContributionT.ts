/**
 * Defines the type alias for a `Contribution` record.
 */
export type Record = {
  amount: number;
  legislatorId: string;
  organization: string;
};
