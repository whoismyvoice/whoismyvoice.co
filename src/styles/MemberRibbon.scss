@import 'variables';

.ribbon-wrapper {
  z-index: 2;
  right: auto;
  top: -32%;
  left: -1.5%;
  position: relative;
  margin: 0 auto;
  width: 65%;
  min-width: 220px;
  transform: rotate(-2deg);
  font-family: 'Open Sans';

  strong {
    font-weight: 600;
  }

  @media screen and (max-width: 1040px) {
    width: 85%;
    top: -15%;
    left: 0;
  }

   @media screen and (max-width: 1170px) {
    width: auto;
    top: -28%;
    left: 0;
  }

  @media screen and (max-width: $iphone6) {
    width: 80%;
    transform: rotate(0deg);
    min-width: 120px;
    top: -20%;
    left: 0;
  }

  @media screen and (max-width: $iphone5-min) {
    width: 100%;
    top: -20%;
  }
}

.ribbon {
  position: relative;
  left: auto;
  right: auto;
  width: 100%;
  margin: 0 auto;
  min-width: 100%;
  padding: 0.4em 0.3em 0.6em 0.3em;
  font-size: 16px;
  text-transform: uppercase;
  background-color: $light-red;
  text-align: center;
  color: #ecf0f1;

  @media screen and (max-width: 1040px) {
    width: 80%;
    min-width: 60%;
    font-size: 14px;
  }

  @media screen and (max-width: 1170px) {
    width: 80%;
    font-size: 14px;
  }

  @media screen and (max-height: 650px) {
    width: 70%;
    min-width: 220px;
  }

  @media only screen
  and (min-device-width: 768px)
  and (max-device-width: 1024px)
  and (orientation: portrait)
  and (-webkit-min-device-pixel-ratio: 1) {
    font-size: 13px;
    width: 65%;
  }

  @media only screen
  and (min-device-width: 768px)
  and (max-device-width: 1024px)
  and (orientation: landscape)
  and (-webkit-min-device-pixel-ratio: 1) {
    font-size: 13px;
    width: 60%;
  }

  @media screen and (max-width: $iphone6) {
    min-width: 100%;
    width: 100%;
    font-size: 9px;
    padding: 0.5em 0.5em 0.5em 0.5em;
  }

   @media screen and (max-width: $iphone5-min) {
    width: 80%;
    min-width: 90%;
    font-size: 9px;
    padding: 0.3em 0.3em 0.3em 0.3em;
  }
}

.ribbon--blue {
  background-color: #3569CA;
}

.ribbon:before, .ribbon:after {
  z-index: -1;
  position: absolute;
  display: block;
  content: "";
  bottom: -0.8em;
  border: 1.5em solid darken(#B03532, 0%);

  @media screen and (max-width: $iphone6) {
    border: 0.8em solid darken(#B03532, 0%);
  }
}

.ribbon--blue:before, .ribbon--blue:after {
  border: 1.5em solid darken(#203C70, 0%);
}

.ribbon:before {
  left: -2em;
  border-top-color: transparent;
  border-left-color: transparent;
  border-top-width: 0px;
  border-left-width: 25px;
  border-bottom-width: 40px;
  border-right-width: 30px;

  @media screen and (max-width: 1040px) {
    border-left-width: 15px;
    border-bottom-width: 40px;
    border-right-width: 30px;
    left: -1.6em;
  }

  @media screen and (max-width: 1170px) {
    border-left-width: 15px;
    border-bottom-width: 40px;
    border-right-width: 30px;
    left: -1.8em;
  }

  @media screen and (max-width: $iphone6) {
    left: -1.8em;
    border-left-width: 12.5px;
    border-bottom-width: 25px;
    border-right-width: 18px;
  }

  @media screen and (max-width: $iphone5-min) {
    left: -1.2em;
    border-left-width: 12px;
    border-bottom-width: 20px;
    border-right-width: 12px;
  }
}

.ribbon:after {
  right: -2em;
  border-top-color: transparent;
  border-right-color: transparent;
  border-top-width: 0px;
  border-right-width: 25px;
  border-bottom-width: 40px;
  border-left-width: 30px;

  @media screen and (max-width: 1040px) {
    border-right-width: 15px;
    border-bottom-width: 40px;
    border-left-width: 30px;
    right: -1.6em;
  }

  @media screen and (max-width: 1170px) {
    border-right-width: 15px;
    border-bottom-width: 40px;
    border-left-width: 30px;
    right: -1.8em;
  }

  @media screen and (max-width: $iphone6) {
    right: -1.8em;
    border-right-width: 12.5px;
    border-bottom-width: 25px;
    border-left-width: 18px;
  }

   @media screen and (max-width: $iphone5-min) {
    right: -1.2em;
    border-right-width: 12px;
    border-bottom-width: 20px;
    border-left-width: 12px;
  }
}

.ribbon .ribbon-content:before, .ribbon .ribbon-content:after {
  position: absolute;
  bottom: -0.8em;
  content: "";
  border-style: solid;
  border-color: darken(#B03532, 8%) transparent transparent transparent;
}

.ribbon .ribbon-content:before {
  left: 0;
  border-width: 0.8em 0 0 1.5em;
}

.ribbon .ribbon-content:after {
  right: 0;
  border-width: 0.8em 1.5em 0 0;
}

.ribbon--blue .ribbon-content:before, .ribbon--blue .ribbon-content:after {
  border-color: darken(#203C70, 8%) transparent transparent transparent;
}

