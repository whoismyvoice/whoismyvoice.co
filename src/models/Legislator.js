// @flow

import PropTypes from 'prop-types';
import { PropType as ContributionType } from './Contribution';

// Types
import type { ContributionRecord } from './Contribution';

export type BioguideId = string;
export type FecId = string;
opaque type LegislatorChamber = 'senate' | 'house' | 'unknown';
export type LegislatorIdentifier = string;

type LegislatorBioRecord = {
  birthday: string,
  gender: string,
};

export type LegislatorChannelRecord = {
  id: string,
  type: string,
};

type LegislatorIdRecord = {
  bioguide: BioguideId,
  fec: Array<FecId>,
};

type LegislatorNameRecord = {
  first?: string,
  last?: string,
  official_full: string,
};

type LegislatorTermRecord = {
  type: string,
  start: string,
  end: string,
  state: string,
  district?: string,
  party: string,
  phone: string,
};

export type LegislatorRecord = {
  id: LegislatorIdRecord,
  identifier?: string,
  name: string | LegislatorNameRecord,
  bio: LegislatorBioRecord,
  channels: Array<LegislatorChannelRecord>,
  photoUrl?: string,
  terms: Array<LegislatorTermRecord>,
};

/**
 * Defines the `PropTypes` validation structure for a `Legislator` record.
 */
export const PropType = PropTypes.shape({
  id: PropTypes.shape({
    bioguide: PropTypes.string,
    fec: PropTypes.arrayOf(PropTypes.string),
  }),
  identifier: PropTypes.string,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      first: PropTypes.string,
      last: PropTypes.string,
      official_full: PropTypes.string.isRequired,
    }),
  ]),
  bio: PropTypes.shape({
    birthday: PropTypes.string,
    gender: PropTypes.string,
  }),
  channels: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      type: PropTypes.string,
    })
  ),
  photoUrl: PropTypes.string,
  terms: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string.isRequired,
      start: PropTypes.string.isRequired,
      end: PropTypes.string.isRequired,
      state: PropTypes.string,
      district: PropTypes.number,
      party: PropTypes.string,
    })
  ),
});

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator {
  /**
   * Get the legislator's identifier from the given `record`.
   * @param {LegislatorRecord} record that matches the `PropType` defined for
   *    `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: LegislatorRecord): LegislatorIdentifier {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator#getIdentifier'
    );
    let identifier = undefined;
    if (record.identifier !== undefined) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else {
      identifier = stripDiacritics(record.name.official_full);
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {LegislatorRecord|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<ContributionRecord>,
    legislator: LegislatorRecord | Legislator
  ): number {
    PropTypes.checkPropTypes(
      { contributions: PropTypes.arrayOf(ContributionType) },
      { contributions },
      'contributions',
      'Legislator#getContributionAmount'
    );
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  record: LegislatorRecord;

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {LegislatorRecord} record containing data about the legislator.
   */
  constructor(record: LegislatorRecord) {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator'
    );
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): LegislatorBioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): LegislatorChamber {
    if (this.isSenator) {
      return 'senate';
    } else if (this.isRepresentative) {
      return 'house';
    } else {
      return 'unknown';
    }
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): LegislatorTermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): LegislatorIdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): LegislatorIdentifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): LegislatorNameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): string {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<LegislatorTermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): ?string {
    const channels = this.record.channels || [];
    const channel = channels.find(channel => channel.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
