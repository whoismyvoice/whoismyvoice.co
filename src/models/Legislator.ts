import { Contribution } from './Contribution';
import { Record as Official } from './Official';

export type TermType = 'sen' | 'rep';
export enum Party {
  Democrat = 'Democrat',
  Republican = 'Republican',
  Independent = 'Independent',
}
export type BioguideId = string;
export type FecId = string;
export enum Chamber {
  Senate,
  House,
  Unknown,
}
export type Identifier = string;

export interface BioRecord {
  birthday: string;
  gender: string;
}

export interface ChannelRecord {
  id: string;
  type: string;
}

export interface IdRecord {
  bioguide: BioguideId;
  fec: Array<FecId>;
}

export interface NameRecord {
  first?: string;
  last?: string;
  official_full: string;
}

export interface TermRecord {
  type: TermType;
  start: string;
  end: string;
  state: string;
  district?: string;
  party: Party;
  phone: string;
}

export interface Record {
  id: IdRecord;
  identifier?: string;
  name: string | NameRecord;
  bio: BioRecord;
  channels: Array<ChannelRecord>;
  photoUrl?: string;
  terms: Array<TermRecord>;
}

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str: string) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator implements Record {
  /**
   * Holds the Record used to retrieve data for getters.
   */
  private record: Record;

  /**
   * Get the legislator's identifier from the given `record`.
   * @param {Record | Official | Legislator} record that matches the `PropType`
   *    defined for `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: Record | Official | Legislator): Identifier {
    let identifier = undefined;
    if (record instanceof Legislator) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else if (typeof record.name !== 'undefined') {
      identifier = stripDiacritics(record.name.official_full);
    } else {
      identifier = record.identifier || 'unknown';
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {Record|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<Contribution>,
    legislator: Record | Legislator
  ): number {
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {Record} record containing data about the legislator.
   */
  constructor(record: Record) {
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): BioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): Chamber {
    if (this.isSenator) {
      return Chamber.Senate;
    } else if (this.isRepresentative) {
      return Chamber.House;
    } else {
      return Chamber.Unknown;
    }
  }

  /**
   * Retrieve the channels from by which the legislator can be reached.
   * @returns an array of channels.
   */
  get channels(): Array<ChannelRecord> {
    return this.record.channels;
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): TermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): IdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): Identifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): NameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): Party {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<TermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): string | undefined {
    const channels = this.record.channels || [];
    const channel = channels.find(c => c.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
