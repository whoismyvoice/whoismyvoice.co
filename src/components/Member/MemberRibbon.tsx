import * as React from 'react';
import * as cx from 'classnames';
import { Legislator, Party } from '../../models/Legislator';

// Styles
import './../../styles/MemberRibbon.css';

interface Props {
  legislator: Legislator;
}

class MemberRibbon extends React.Component<Props> {
  render() {
    const { legislator } = this.props;
    const { fullName, party, state } = legislator;

    const ribbonClasses = cx(
      ['ribbon'],
      { 'ribbon--grey': party === Party.Independent },
      { 'ribbon--blue': party === Party.Democrat }
    );

    return (
      <div className="ribbon-wrapper">
        <h1 className={ribbonClasses}>
          <strong className="ribbon-content">
            ({party.charAt(0)} - {state}) {fullName}
          </strong>
        </h1>
      </div>
    );
  }
}

export default MemberRibbon;
