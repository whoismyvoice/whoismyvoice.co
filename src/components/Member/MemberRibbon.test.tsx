import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as renderer from 'react-test-renderer';
import MemberRibbon from './MemberRibbon';
import { createLegislator } from '../../models/Legislator.test';
import { Legislator } from '../../models/Legislator';

jest.mock('mixpanel-browser');

const props = {
  legislator: new Legislator(createLegislator('John Smith')),
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MemberRibbon {...props} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders correctly', () => {
  const tree = renderer.create(<MemberRibbon {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
