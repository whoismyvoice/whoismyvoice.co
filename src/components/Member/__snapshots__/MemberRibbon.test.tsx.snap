// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ribbon-wrapper"
>
  <h1
    className="ribbon ribbon--grey"
  >
    <strong
      className="ribbon-content"
    >
      (
      I
       - 
      NY
      ) 
      John Smith
    </strong>
  </h1>
</div>
`;
